File: Data Structures & Algorithms/longest-increasing-subsequence/submission-2.cpp
class Solution {
public:
    int lengthOfLIS(vector<int>& nums) {
        vector<vector<int>> cache(nums.size(), vector<int>(nums.size() + 1, -1));
        return ans(nums, 0, -1, cache);
    }

    int ans(vector<int>& nums, int index, int prevIndex, vector<vector<int>>& cache) {
        if (index >= nums.size()) {
            return 0;
        }
        if (cache[index][prevIndex + 1] != -1) {
            return cache[index][prevIndex + 1];
        }
        
        int ifNotPicked = ans(nums, index + 1, prevIndex, cache);
        int ifPicked = -1;
        if ((prevIndex == -1) || (nums[index] > nums[prevIndex])) {
            ifPicked = 1 + ans(nums, index + 1, index, cache);
        }
        
        cache[index][prevIndex + 1] = max(ifPicked, ifNotPicked);
        return cache[index][prevIndex + 1];
    }
};

File: Data Structures & Algorithms/coin-change/submission-2.java
class Solution {
    public int coinChange(int[] coins, int amount) {
        Map<Integer, Integer> dp = new HashMap<>();
        int ans = dfs(coins, amount, 0, dp);
        return ans >= (int)1e9 ? -1 : ans;
    }

    private int dfs(int[] coins, int amount, int index, Map<Integer, Integer> dp) {
        if (amount == 0) {
            return 0;
        }
        if (index == coins.length) {
            return (int)1e9;
        }
        if (dp.containsKey(amount)) {
            dp.get(amount);
        }

        int ifPicked = (int)1e9, ifNotPicked = dfs(coins, amount, index + 1, dp);
        if (coins[index] <= amount) {
            ifPicked = 1 + dfs(coins, amount - coins[index], index, dp);
        }

        dp.put(amount, Math.min(ifPicked, ifNotPicked));
        return dp.get(amount);
    }
}
